### ISSUE TITLE
Build Checkout panel

### ISSUE THREAD

**Issue by NullOperator404** [NullOperator404/Developer-Diary#34]
9/28/24

The majority of the Checkout panel is finished, except for several key issues: 

1. 
The PayPal/Debit card transaction section still does not work properly. Unfortunately, this section is difficult to troubleshoot since the code is an insert from PayPal, and their own site's instructions on how to implement it leaves much to be desired. Obviously, the idea would be to perform mock transactions using a mock credit card # provided by PP. It's been nearly a year since I've worked on this, but if memory serves, one of the issues I was facing was getting the final dollar amount of the sale transferred to the PP applet so that it would perform the transaction for that amount. I will need to review it further and in more detail to refresh my memory; I may have resolved this problem last year, as there's no mention of it in any of my other entries on GitHub. 

2. 
Like everything else in the lab, the way this panel has been designed assumes the user is on a PC or MAC (the panel slides into view from the right). This of course, won't work if the user is on a mobile device, so in effect the entire layout of the panel will need to be reformatted using a separate CSS file. However, after a quick glance, I don't think it would be much of a problem - the information could just be displayed in a "stacked" layout, with the "proofs" at the top, then rows for quantities, then promo codes, then subtotals/totals box, then the PP buttons to complete the transaction. This way, the panel could still slide in from the right, but would fit on a vertical aspect ratio.

3. There is a minor bug that needs to be addressed regarding deleted items. See the "BUG" item for the project.

**Comment by NullOperator404** [comment]
UPDATE 6/10/25

I had to put this project on hold for some time since the fall of last year after securing a web design/dev project for a client (PBs BBQ). 

While working on that project, which was also an e-commerce site, I learned a much simpler way to manage shopping cart contents and relay the data to a separate Checkout page. Because of this, I want to refactor the code in the Design Lab to the PB's model, which is much easier to manage. Here's a breakdown:

1. An Object Array should be created in the JS file for a "Shopping Cart." This array will store all pertinent data for a given product. If the user adds additional products, additional objects will be added to the array. The PreFlight & Checkout screens both will populate themselves based on data stored in this array, saving the device from having to rely on real-time data from the DOM. 

2. The object array (let's call it `shoppingCart` ) should be formatted as follows:

```
let shoppingCart = [
{
     "ProductNumber": "G500",
     "ProductColor": "White",
     "ProductDescription": "yadda yadda yadda",
     "Sizes": (import data from JSON database),
     "Quantities": [5, 5, 5, 5, 5],
     "BasePrice": 25.35,
     "DesignProofs": Null*
}
];
```

*All values shown here are defaults; index 0 will update if the user selects a different product from the design space. Values shown as Null will not be assigned a value until the user clicks the "Get Quote" button. `DesignProofs` will be populated with the HTML capture used to display the proofs in the PreFlight window. 

The `Quantities` array will be populated when the user clicks the Checkout button, and will pull data from the input boxes. Ascending indices will correspond to left-to-right reading of the sizes. (Does that make sense?)

Adding products to the order from the PreFlight screen will add additional objects to the `shoppingCart` array. 

Ultimately, the goal here is to compartmentalize all order data in one place so that the Checkout screen need only read this array's contents to summarize the customer's order. Likewise, the cart data could be used to generate a work order to be emailed to me upon order submission. This keeps the process neat and tidy. 

3. Benefits to this change

- If the user inputs quantities and then returns to the design space, initial order inputs can now be preserved and displayed when the user returns to PreFlight mode (pulling from `Quantities` to populate the boxes).
- Checkout can now be moved to a separate page entirely via `SessionStorage`, which will store the `shoppingCart` data. 
- Cart contents can now be saved and preserved across all pages on the website, in case the user wishes to leave the design lab and return later (even the lab itself could be "restored" to their existing design by pulling data from the cart). 
- Deleting items from the cart on the PreFlight screen no longer requires a convoluted system of tracking rows, boxes, etc; each delete icon can be given a unique ID, and this can be used to determine which item to delete from the `shoppingCart` array. Upon deletion, the rows can then be "redrawn" based on the cart data and boxes will retain input data. 

4. The following functions will need to access and update the `shoppingCart`:

1. `updateNums()`
2. `addToCart()`
3. `openCheckout()`
4. `trashItem()`
5. `openPreFlight()`
6. `selectProd()`

**Comment by NullOperator404** [comment]
UPDATE 6/15/2025

The new shopping cart has been implemented, but as expected, it has resulted in a cascade of bugs & errors. Most of these problems stem from the application being so intricate and complex with numerous functions depending on everything working in tandem. The good news is that I've slowly worked through most of the bugs/errors and the Preflight process is stabilized again, this time operating under the new code. 

The `shoppingCart[]` array has allowed me to compartmentalize several other blocks of code. 

1. A new function called `generateCartContents()` handles drawing the cart contents on the Preflight screen. Moving this code to it's own function allowed repetitive blocks to be eliminated and replaced with a one-line function call (anytime the cart needs to be redrawn). 

2. I wrote a new function called `calculateTotals()` tonight, which has eliminated the need for `trashItem()` to be cluttered with code copied from `updateNums()`; when the user deletes an item, the cart totals must be recalculated of course. 

3. The new shopping cart approach has also allowed inputted quantities to be saved even if the user exits Preflight mode. Upon re-opening Preflight, JS will check the `shoppingCart[]` array to see if any inputted quantities have been recorded, and if so, it will insert those values into their respective quantity boxes. This means that no matter the reason for the cart contents needing to be redrawn, any previously inputted quantities will persist. Eventually, I'll add code to synchronize `sessionStorage[]` with the `shoppingCart[]`, meaning that even if the user exits the design lab entirely after inputting quantities and returns later, JS will be able to remember their inputs by re-populating `shoppingCart[]` by pulling from `sessionStorage`. 

----

The bug I'm currently stuck on is that the math is totally whack -- that is, the Total Pieces does not reflect correct quantities; keying in "12" requires pressing "1" then "2," so I think JS is first reading "1," then reading "12," and adding them together (13). Even REMOVING numbers causes the Total Pieces to INCREASE. 

Since the total dollar amount is calculated based on quantities, it too, is totally incorrect. 

Unfortunately, problems like these are difficult to troubleshoot because _technically_, the application is working as it's being told to do. (No errors in the console). So the problem is being caused by a breakdown in logic somewhere. 

Once this issue is fixed (and assuming no others crop up because of it), I want to focus on adding a "cart icon" at the top similar to the PB's website. The icon will appear at the top of all pages, and when clicked, will open the Preflight screen (this will require adding a ton of HTML & CSS to the other pages, as they will all need to be able to support the Preflight screen....I dunno....the more I think about it, the more daunting it seems; A TON of Javascript is needed to make that thing work).

**Comment by NullOperator404** [comment]
ADDENDUM:

I really want to see if there's someway I can simplify the process further so that it doesn't depend on so many arrays, variables, etc all being synchronized. I feel like a lot of this "monitored data" could be stored in an object array, so rather than have a bunch of arrays & vars peppered throughout the code, just initialize a global object array in one place and have the functions update their respective data in that array. 

For example, total number of pieces, total dollar amount, row totals, row prices, etc...just to name a few. These vars are accessed and modified all over the place, and it would make debugging a lot faster if I could just `console.log()` the array and see at a glance what each of the values are.

**Comment by NullOperator404** [comment]
UPDATE 6/16/2025

The math error has been resolved. Very easy fix. What happened was that in order to make `calculateTotals()` work, most of the vars in the `updateNums()` function had to be moved to global so `calculateTotals()` could access them. When the vars were originally initialized locally, they were set with a value of zero, meaning that each time `updateNums()` ran, the vars were reset to zero since the function calculates totals by looping through the quantity boxes. When I moved the vars to global, they were not being reset to zero each time the function ran, so keying in "5" under Small and then keying in "5" under Medium would result in "15," 

Two lines resetting the vars within `updateNums()` resolved the issue:

```
rowTotal = 0;
rowPrice = 0;
```

**Comment by NullOperator404** [comment]
UPDATE 6/22/2025

Refactoring of the Checkout page is now virtually complete. I'll want to run a few more tests to be sure, but everything seems to be running smoothly. Several objects had to be added to the shopping cart object-array to store user selections such as names, numbers, name/number sizes, which side of the garment they go on, and any notes they typed in. All of this information is displayed for the user on the checkout page. 

When the user clicks "Checkout" on the Preflight panel, JS creates a new object and pushes it into the `shoppingCart[]` array. This object contains key-value pairs storing dollar data - subtotal, tax, shipping, etc (all of these figures must be determined during Preflight, since that's where the user is inputting their order). `shoppingCart[]` is then copied to `sessionStorage[]` for transfer to the Checkout page. 

The code has been written such that the "dollar values" object will ALWAYS be the last object in the array. 

To summarize flow: 
- User opens Preflight by clicking "Get Quote" 
- User inputs quantities, but only the Preflight panel is updated in real-time. Adding products from the Preflight panel will push additional objects into the `shoppingCart[]` array. 
- Deleting an item from the Preflight panel will set that item's index in the array to `null`. All instances where the cart must be looped through check for indices set to `null`, and if they exist, JS ignores them. 
- When the user clicks "Checkout," JS will record all quantities entered and store them in their appropriate slots in the `shoppingCart[]` array. All other pertinent data is passed to the `shoppingCart[]` array - design proof HTML, item description, item color, product number (for work order purposes), saved notes (if any), whether or not customer has added custom names/numbers, a breakdown of all inputted names/numbers, indicating which size garment gets what name, what number, and which side the art goes on, and final dollar amounts. The user is then redirected to the Checkout page. 
- When the Checkout page loads, it reads `sessionStorage` and retrieves the data, then passes it into a new var. All pertinent order data is then outputted to the Checkout page so the user can review before making payment. 

All that remains (barring further QA testing) is to finish setting up the PayPal payment portal, which will no doubt be a fustercluck since it's primarily back-end based. Ugh. 

For now, I'm going to work on eliminating some of the smaller bugs I have logged here on GitHub.

**Comment by NullOperator404** [comment]
UPDATE 4/3/2026

Having spent last year working on a checkout pipeline for the PB's BBQ project, I've gained a much better understanding of backend flow and after examining my current setup, I now realize this one won't work. 

The method I currently have implemented is PayPal's "Pay Links & Checkout Buttons," which is essentially a code-free "insert our link here" method. This method only works for fixed-priced products, and isn't compatible with a shopping cart application since the cart total varies from sale to sale. 

My use-case is the same as PB's BBQ, in which PayPal would need to receive order total info from the frontend. Additionally, since PayPal would only be handling payment(s), logic must be added to do the following:

1. A email-friendly order summary/invoice/receipt must be dynamically built and then sent to the customer's email after the transaction is complete. Building it won't be too difficult, but packaging this into an HTML document and programmatically emailing it to the customer will require backend stuff I'm unfamiliar with. 

2. All pertinent order data will need to be "packaged" and sent to MY email -- design proofs, notes, quantities, sizes, names/numbers, etc -- order $ totals should be included as well for good measure. Customer name/shipping info will need to be sent along as well.

**Comment by NullOperator404** [comment]
UPDATE 4/3/2026 (20 minutes later)

I may have been mistaken in my previous entry -- it seems that what I actually have implemented is PayPal's SDK, not the "drop link." The Javascript SDK allows for total control over payment amounts, order info, etc. So what I think I need to do now is:

1. Create a server-side JS file that will contain the following:

a) A call to create an order via PayPal:
FRONT-END:
```
paypal.Buttons({
                async createOrder() {
                    const response = await fetch("/my-server/create-paypal-order", {
                        method: "POST",
                        headers: {
                            "Content-Type": "application/json",
                        },
                        body: JSON.stringify({
                            cart: [{
                                sku: "YOUR_PRODUCT_STOCK_KEEPING_UNIT",
                                quantity: "YOUR_PRODUCT_QUANTITY",
                            }]
                        })
                    });

                    const order = await response.json();

                    return order.id;
                }
            }).render('#paypal-button-container');
```

BACKEND (NODE.JS):
```
app.post("/my-server/create-paypal-order", async (req, res) => {
  const order = await createOrder();
  res.json(order);
});

// use the orders api to create an order
function createOrder() {
  // create accessToken using your clientID and clientSecret
  // for the full stack example, please see the Standard Integration guide
  // https://developer.paypal.com/docs/multiparty/checkout/standard/integrate/
  const accessToken = "REPLACE_WITH_YOUR_ACCESS_TOKEN";
  return fetch ("https://api-m.sandbox.paypal.com/v2/checkout/orders", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${accessToken}`,
    }
    body: JSON.stringify({
      "purchase_units": [
        {
          "amount": {
            "currency_code": "USD",
            "value": "100.00"
          },
          "reference_id": "d9f80740-38f0-11e8-b467-0ed5f89f718b"
        }
      ],
      "intent": "CAPTURE",
      "payment_source": {
        "paypal": {
          "experience_context": {
            "payment_method_preference": "IMMEDIATE_PAYMENT_REQUIRED",
            "payment_method_selected": "PAYPAL",
            "brand_name": "EXAMPLE INC",
            "locale": "en-US",
            "landing_page": "LOGIN",
            "shipping_preference": "GET_FROM_FILE",
            "user_action": "PAY_NOW",
            "return_url": "https://example.com/returnUrl",
            "cancel_url": "https://example.com/cancelUrl"
          }
        }
      }
    })
  })
  .then((response) => response.json());
}
```
The backend file should also include a call to capture payment:
```
async onApprove(data) {
        // Capture the funds from the transaction.
        const response = await fetch("/my-server/capture-paypal-order", {
          method: "POST",
          body: JSON.stringify({
            orderID: data.orderID
          })
        })

        const details = await response.json();

        // Show success message to buyer
        alert(`Transaction completed by ${details.payer.name.given_name}`);
    }
}).render('#paypal-button-container');
```

Unfortunately, much like Clover systems, PayPal has a predefined JSON structure that their servers will expect, so my "order data" will need to be reformatted to accommodate what PayPal expects. I'd still like to include a function that does what I outlined above -- sending a customized-to-my-specifications order summary to myself and to the customer.

**Comment by NullOperator404** [comment]
UPDATE 4/3/2026 (30 MINS LATER)

After further review, it looks like most of the above code was already in place -- I already had a backend file `server.js` setup. A couple of log statements at the start of `app.js` confirmed the dollar total and quantity total were being read from `checkout.js`. 

I started running `server.js` in the Command Prompt and after clicking the PayPal button, I got a CORS error (of course). I do recall encountering this issue last year when I was working on it, but at the time didn't fully understand what was happening. 

Thankfully, I encountered this same problem on the PB's BBQ project and learned how to resolve it. The following line of code is added to the top of `server.js`:

`import cors from "cors";`

Along with this:

```
app.use(cors({
  origin: "*", // You can restrict this to your site later (e.g. "http://127.0.0.1:5500")
  methods: ["GET", "POST", "OPTIONS"],
  allowedHeaders: ["Content-Type", "Authorization"]
}));
app.use(express.json());
```

^ This block of code bypasses CORS and allows my browser (PORT 63452) to talk to my server (PORT 3000). 

This simple implementation copied from the PB's server file allowed it to work -- clicking the PayPal button now opens the PayPal payment window. It's currently opening in Sandbox mode, and a predefined amount of $100 is showing. I need to see how/why the amount is $100, and I need to start trying to run test transactions. 

ADDENDUM:
Clicking the "Pay with Debit/Credit" button expands a form where the customer can enter shipping & card info, then click "Pay"

I suspect that when I set up my developer account with PayPal, I probably set the Sandbox amount to $100. I'm going to check now.

**Comment by NullOperator404** [comment]
UPDATE 4/4/2026 (12:02 AM)

I logged into my dev account and found my "test card" credentials, then attempted a transaction. 

I am getting an error in the browser console from PayPal:

`onApprove_non_resume_flow_merchant_callback_rejected`
` Cannot read properties of undefined (reading '0')`

I'm not sure where exactly the error is coming from, or where it's reading a value of '0'. No errors in the server console.

**Comment by NullOperator404** [comment]
UPDATE 4/6/2026

After consulting with ChatGPT, the error was the result of a simple syntax error in the `onApprove` call - the endpoint being called was `create-paypal-order` when it should be `capture-paypal-order`. This calls the correct endpoint to confirm payment. Two tests both went through successfully using my sandbox credit card. 

A few things are needed now:

1. Setup a redirect "Thank you for your order!" page, which should have a button "Done" or "OK," which when clicked will return the user to the Dynamik Graphics home page. 

2. A order-numbering system should be implemented similar to the one used for PB's BBQ, which randomly generates a unique number and assigns it to the order -- this will insure that both the customer and me can reference a specific order without relying on art descriptions, dates, etc. 

3. A JSON file should be setup that stores customer info, including design details. Later I can implement a recall system that would allow users to load old designs. This isn't essential at the moment, but should eventually be done.

4. A system needs to be built that will generate an email-friendly order confirmation/order summary that will be sent to the customer, as well as a copy sent to MY email for order processing. To start with, I can setup a simple "send an email to me" function that will send me an email upon an order completion.

**Comment by NullOperator404** [comment]
UPDATE 4/13/2026

The redirect "Thank you for your order" page is virtually finished. Upon a successful transaction, users will be redirected to this page that displays their order number, order total, total number of pieces ordered, and an estimated ship date. A "Done" button returns them to the website landing page. I want to add a line of text that notifies them that a confirmation email will be sent to them. 

An order numbering system has also been implemented in `checkout.js`, which generates a random number between 1 - 10,000. If/when business grows, this system may need to be changed to six digits, but it's unlikely. 

For the email side of things, I plan to utilize code written for PB's BBQ, then modify it. The following block of code should be added to `redirect.js`: 

```
document.getElementById('contactForm').addEventListener('submit', async function (e) {
//THE ABOVE LINE OF CODE SHOULD BE CHANGED SO THAT THE FUNCTION FIRES WHEN THE PAGE LOADS, NOT WHEN A FORM IS SUBMITTED
  e.preventDefault(); // prevent default form reload

//FOR NOW, A BASIC STRING WILL BE USED AS TEST DATA
 const data = "Test data to later be  changed to HTML & CSS email format"

  try {
    const response = await fetch('http://localhost:3000/contact-submit', { //THIS WILL NEED TO CALL A SERVER FILE THAT HANDLES ORDER CONFIRMATION EMAILS
      method: 'POST',
//THE CONTENT TYPE NEEDS TO BE CHANGED TO TEXT; JSON WON'T BE NECESSARY.
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(data),
    });

    const result = await response.json();

    if (result.success) {
      //FOR NOW, MAYBE AN ALERT MESSAGE CONFIRMING SUCCESS?
    } else {
      alert('Error: ' + result.message);
    }
  } catch (error) {
    console.error('Submission failed:', error);
    alert('Failed to send message.');
  }
});
```

The server file should look something like this. Nodemailer.js will need to be installed. 
```
const express = require('express');
const bodyParser = require('body-parser');
const nodemailer = require('nodemailer');
const cors = require('cors');

const app = express();
const port = 3000;

// Middleware
app.use(cors());
app.use(bodyParser.urlencoded({ extended: false }));
app.use(bodyParser.json());

// POST endpoint to handle form submission
app.post('/contact-submit', async (req, res) => {
//THE VAR BELOW NEEDS TO BE CHANGED SO THAT IT EXPECTS ONLY TEXT FROM THE BODY, NOT JSON DATA
  const { fname, lname, phoneNum, email, message } = req.body;

  // Configure your email transport
  let transporter = nodemailer.createTransport({
    host: 'smtp.office.com',
    port: 465,
    secure: true,
    service: 'Gmail', //THIS WILL NEED TO BE CHANGED TO MY SERVICE PROVIDER. OFFICE? MICROSOFT? I NEED TO CONFIRM
    auth: {
      user: 'art@dynamikgraphics.com',
      pass: 'Dynamik82!',
    },
  });

  let mailOptions = {
    from: 'The Website',
    to: 'art@dynamikgraphics.com',
    subject: "Message from Dynamik Graphics",
    text: message, //IN THIS CASE, THE VAR 'MESSAGE' WILL CONTAIN THE TEXT DATA
  };

  try {
    await transporter.sendMail(mailOptions);
    res.status(200).json({ success: true, message: 'Message sent successfully!' });
  } catch (error) {
    console.error('Error sending email:', error);
    res.status(500).json({ success: false, message: 'Failed to send message.', error: error.message });
  }
});

app.listen(port, () => {
  console.log(`Server running on http://localhost:${port}`);
});
```

**Comment by NullOperator404** [comment]
UPDATE 4/14/2026

I have implemented the above code and like the tests I ran with PBs, the confirmation page now automatically sends an email when it loads. However, there are a few problems:

1. My work email is hosted by Microsoft Office 365, and they have much tighter backend security than Yahoo! or Gmail, so unfortunately connecting to that account isn't going to be as straightforward as connecting to my Yahoo account (which I used for these tests). ChatGPT recommends a third-party API called Resend, which will work similarly to CloudConvert. I will need to set this up, as I don't want online orders going to my Yahoo email; I rarely check it. 

2. At present, the test email uses a simple string of text, but this will eventually need to be formatted into an HTML/CSS email. Inserting HTML tags into the current setup simply treats them as part of the "message," so the HTML is not rendered on-screen when the email is viewed. Also, I could only make this work by setting the `Content-Type` to `application/json` and formatting the data as JSON. This is normal for using `<form>` data, but since this just needs to be an email formatted in HTML/CSS to display the order summary, the `Content-Type` will eventually need to be changed to `text/html`. 

3. Shouldn't be too hard, but when the email is fired, code needs to be written so that TWO copies are sent out - one to MY email, and the other to the customer's email, which can be pulled from the PayPal payment form (I think).

**Comment by NullOperator404** [comment]
UPDATE 4/16/2026

Success! 

Resend was extremely easy to implement. The following code is dropped into my `async /confirm-order` function:

```
const resend = new Resend('MY-API-KEY-INSERTED-HERE');
  console.log(req.body);

      resend.emails.send({
      from: 'onboarding@resend.dev',
      to: 'art@dynamikgraphics.com',
      subject: 'Hello World',
      html: '<p>Congrats on sending your <strong>first email</strong>!</p>'
    });
```

All I need to do now is modify this code (provided by Resend) to my specifications:

1. "From" needs to be changed to "Dynamik Graphics"

2. "To" needs to stay as it is, although a second function needs to be written where "To" is  a var containing the user's email address.

3. "Subject" should be changed to "Thank you for your order!" or "Order confirmation." 

4. "HTML" -- self-explanatory. I now need to build/format an HTML email for delivery purposes.

**Comment by NullOperator404** [comment]
UPDATE 4/17/2025

The above modifications have been made and preliminary work has begun on the HTML email template that will be used. To make things easier, I copied the code from my promo email and am using it as a general framework for the order confirmation email - this way, I can work on the HTML in a separate window in IntelliJ and view the email in the browser. Once the template is ready, I can just copy the code stack and paste into a var I've setup in `confirm-order.js` called `html`, which will contain the HTML/CSS of the email. The var will then be set as the value of `html` in the email structure. 

A few minor things are needed now:
The confirmation email needs to display the proofs the customer views at checkout, so all design data will need to be imported to `confirm-order.js` via the cart data (`sessionStorage`). I _think_ I already have this implemented, but I need to check the code. 
Likewise, all order data will need to be retrieved including sizes, quantities, descriptions, etc...pretty much all info that appears in the checkout page. 

Like the pre-flight & checkout pages, the order summary section in the email will need to dynamically generate rows by iterating through the shopping cart.

**Comment by NullOperator404** [comment]
UPDATE 4/17/2026 (1 hour later)

After addressing the issue of getting cart data to the email template, I realized that the cart data - which exists in `sessionStorage` - can't be read by `confirm-order.js`, since that file is server-side and the cart data exists on the client-side. 

So the email will need to be built client-side in `redirect.js`, then the entire HTML/CSS stack passed as a JSON string to `confirm-order.js` to be inserted into the email body.

**Comment by NullOperator404** [comment]
UPDATE 4/17/2026 (2 hours later)

I'm now at a point where I must go through the entire checkout process to test the data flow, and this has presented a new problem - the email & PayPal code being in separate server files means I can only have ONE of them running at any given time in the console, so it occurs to me that all server functions should be consolidated into a single file. So I will be adding the email code to `server.js` and deleting `redirect.js`.

**Comment by NullOperator404** [comment]
UPDATE 4/21/2026

Several things are now working properly:

*Order number* - I forgot that `sessionStorage` can't be directly modified, but rather must be entirely _overwritten_ when an update is needed. Since `redirect.js` would need to read the entire contents, break it down, update, and then re-package it to be stored, I found it would be much easier for the order number to be generated in `scripts.js` (the JS handling the lab itself) so that it gets assigned when `sessionStorage` is initially set. This way, `redirect.js` only needs to read the data. 

This approach works fine - order numbers are generated during the lab instance, and are outputted on the "Thank You" page, as well as in the confirmation emails. Now both myself and the customer will have a copy of the order number associated with their order. 

*Confirmation Emails* - a function called `buildEmail()` in `redirect.js` now dynamically generates the confirmation email contents, assigns it to a var `codeStack`, and then sends the data to `server.js` where Resend can use the HTML data to format the email. 
All dynamic content is displaying correctly - Order Number, Product Description, Product Color, and a breakdown of sizes/quantities similar to the checkout page. I want to add a table section below that displays the dollar data - Subtotal, Discount (if applicable), Tax, Shipping, & Grand Total. The estimated ship date generated on the Thank You page should also be passed to the email. 

A few minor tweaks to the email body, and then all that's left on this front is setting up a block of code that sends a duplicate email to the customer. The customer's email address needs to be captured from the checkout page and passed to the server for email use. Not sure yet how to do this, as PayPal handles the form data where the email address is entered, so a way to piggyback off either `app.js` or `paypal-api.js` will need to be implemented. 

NOTE TO SELF: 
I need to go through the files and delete any dead-end files that were created when I initially worked on integrating PayPal. There are several JS files that serve no purpose now; as far as I can tell:

`checkout.html` has only two JS files embedded at the bottom: `checkout.js` (which handles checkout functions) and `app.js`, which contains the code sourced from PayPal that handles the checkout buttons, forms, etc. Fetch calls in `app.js` refer to async functions in `server.js`, so I'm not sure what the following files do at this point, as they were not referenced at any point along the pipeline:

paypal-api.js
paypalNode.js
src.js (this file is empty..WTF?)
test.php
testPHP.php

**Comment by NullOperator404** [comment]
UPDATE 4/22/2026

I've styled & added a section for the dollar totals, which of course has presented a new problem. If the customer has a discount code and applies it at checkout, the totals displayed in the DOM are re-calculated, but NOT the totals stored in `sessionStorage`, so the `redirect.html` page has no way of knowing what the discount amount was. The solution to this is tedious, but necessary:

In `checkout.js`, a block of code will need to be added to `promoCodes()` that will assign a new value to a key-value pair in `parseCart`. Since this var contains a JSON version of the cart contents, it should be possible to modify values at this stage. The function will then need to perform a `setItem('CartContents:', parseCart)` function to update `sessionStorage`. The idea is that when the `CartContents` object is initially generated in `scripts.js`, a key-value pair for `Discount` is added with a value of `null`, to be later accessed at checkout. 

My only concern is that I _think_ `setItem` only writes to storage and can't OVERWRITE, meaning a second `CartContents` object would be added to storage at index 2. 

Two additional sections need to be added to the email body that display if the user has added custom names/numbers and/or notes.

**Comment by NullOperator404** [comment]
UPDATE 4/25/2026

In `checkout.js`, a key-value pair has been added to the `shoppingCart` array for `Discount`. By default, this has a value of `null`. 

If a code is applied at checkout, JS now updates this key-value with the savings amount, along with the updated tax & grand total amounts. The `sessionStorage` is cleared, then rewritten using the updated contents of `parseCart` (the JSON parsed version of `sessionStorage`). 

A table has been setup in the email body to display the dollar amounts, and everything is translating correctly over to `buildEmail()`. 

All that remains is adding two table sections in the email body to display customer notes and/or names/numbers. An `in-else` statement should determine whether these sections appear at all. 

Once that's complete, the checkout/transaction pipeline will be completed, and I'll likely move on to styling a mobile-friendly version.

**Comment by NullOperator404** [comment]
UPDATE 4/25/2026 (8 hours later)

Table sections have been added to display customer notes and names/numbers. Conditional logic has been implemented that will only display these sections if there is data present to be displayed. If not, JS will just ignore it. 

The only issue remaining now is the proof pics of the artwork. The background images aren't displaying because the email needs to retrieve them from the server, not from my local drive -- and I haven't added image data to the server yet. However, I am puzzled as to why SVG data isn't at least displaying. This is probably a minor fix. 

I'm fairly happy now with the email side of things, so I'm going to setup a block of code to handle the duplicate for the customer and run a few tests using one of my alternate email accounts. Also, one difference that needs to be added to MY VERSION of the email is that the product number needs to be included (stored in `parseCart`)
